modules/phase_field/src/materials/ElectrochemicalSinteringMaterial.C

```diff
@@ -0,0 +1,345 @@
+
+      // Calculate derivatives of grand potential wrt order parameters and mixed derivatives
+      for (unsigned int i = 0; i < _neta; ++i)
+      {
+        Real sum_domegasdeta = 0.0;
+        for (unsigned int j = 0; j < _ndefects; ++j)
+        {
+          sum_domegasdeta +=
+              -(*_dns_min[j][i])[_qp] * ((*_w[j])[_qp] - _z[j] * _e * _v[_qp] * _v_scale);
+          (*_d2omegasdwdeta[j][i])[_qp] = -(*_dns_min[j][i])[_qp];
+        }
+        (*_domegasdeta[i])[_qp] = sum_domegasdeta;
+
+        for (unsigned int j = i; j < _neta; ++j)
+        {
+          Real sum_d2omegadeta2 = 0.0;
+          for (unsigned int k = 0; k < _ndefects; ++k)
+            sum_d2omegadeta2 +=
+                -(*_d2ns_min[k][i][j])[_qp] * ((*_w[k])[_qp] - _z[k] * _e * _v[_qp] * _v_scale);
+
+          (*_d2omegasdetadeta[i][j])[_qp] = (*_d2omegasdetadeta[j][i])[_qp] = sum_d2omegadeta2;
+        }
+      }
+      break;
+    }       // case 0; // PARABOLIC
+    case 1: // DILUTE
+    {
+      // Calculate grand potential of solid phase and derivatives wrt chemical potentials
+      Real sum_omega_s = 0.0;
+      for (unsigned int i = 0; i < _ndefects; ++i)
+      {
+        Real n_exp = std::exp(((*_w[i])[_qp] - _z[i] * _e * _v[_qp] * _v_scale) / _kB / _temp[_qp]);
+        sum_omega_s += -_kB * _temp[_qp] * (*_ns_min[i])[_qp] * n_exp;
+        (*_domegasdw[i])[_qp] = -(*_ns_min[i])[_qp] * n_exp;
+        (*_d2omegasdw2[i])[_qp] = -(*_ns_min[i])[_qp] * n_exp / _kB / _temp[_qp];
+      }
+
+      _omegas[_qp] = sum_omega_s - 0.5 * _eps_r * eps_0 * _grad_V[_qp] * _grad_V[_qp];
+
+      // Calculate derivatives of grand potential wrt order parameters and mixed derivatives
+      for (unsigned int i = 0; i < _neta; ++i)
+      {
+        Real sum_domegasdeta = 0.0;
+        for (unsigned int j = 0; j < _ndefects; ++j)
+        {
+          Real n_exp =
+              std::exp(((*_w[j])[_qp] - _z[j] * _e * _v[_qp] * _v_scale) / _kB / _temp[_qp]);
+          sum_domegasdeta += -_kB / _temp[_qp] * (*_dns_min[j][i])[_qp] * n_exp;
+          (*_d2omegasdwdeta[j][i])[_qp] = -(*_dns_min[j][i])[_qp] * n_exp;
+        }
+        (*_domegasdeta[i])[_qp] = sum_domegasdeta;
+
+        for (unsigned int j = i; j < _neta; ++j)
+        {
+          Real sum_d2omegadeta2 = 0.0;
+          for (unsigned int k = 0; k < _ndefects; ++k)
+          {
+            Real n_exp =
+                std::exp(((*_w[k])[_qp] - _z[k] * _e * _v[_qp] * _v_scale) / _kB / _temp[_qp]);
+            sum_d2omegadeta2 += -_kB / _temp[_qp] * (*_d2ns_min[k][i][j])[_qp] * n_exp;
+          }
+
+          (*_d2omegasdetadeta[i][j])[_qp] = (*_d2omegasdetadeta[j][i])[_qp] = sum_d2omegadeta2;
+        }
+      }
+
+      break;
+    }       // case 1: // DILUTE
+    case 2: // IDEAL
+    {
+      mooseError(
+          "Ideal solution in solid is not yet supported in ElectrochemicalSinteringMaterial");
+      break;
```

Comment:
Any reason to add this option but not support it?

Reply:
We have a similar model that does not include electrochemical effects, that does have the ideal solution model implemented. I wanted to leave this in place as a more descriptive error in case there are some users that try to move input files from the non-electrochemical model over so they would understand better why the ideal solution option isn't available.

Reply:
I got it and it makes sense.

Would it be better to move this error into the constructor? And change it to ParamError?

Reply:
I'd like to implement the ideal solution case eventually- I think leaving it in the switch statement here leaves the door open to do that eventually more easily